boot.go

```diff
@@ -0,0 +1,113 @@
+package main
+
+import (
+	"log"
+	"os"
+	"time"
+
+	"github.com/deis/registry-token-refresher/pkg"
+	"github.com/deis/registry-token-refresher/pkg/credentials"
+	"k8s.io/kubernetes/pkg/api"
+	kcl "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/labels"
+)
+
+var appList []string
+var registryLocation = os.Getenv("DEIS_REGISTRY_LOCATION")
+
+func getDiff(namespaceList []api.Namespace) []string {
+	var added []string
+	var apps []string
+	// create a set of app names
+	appsSet := make(map[string]struct{})
+	for _, app := range appList {
+		appsSet[app] = struct{}{}
+	}
+
+	for _, ns := range namespaceList {
+		if _, ok := appsSet[ns.Name]; !ok {
+			added = append(added, ns.Name)
+		}
+		apps = append(apps, ns.Name)
+	}
+
+	appList = apps
+	return added
+}
+
+func tokenRefresher(client kcl.SecretsNamespacer, credsProvider credentials.DockerCredProvider, appCh <-chan string, errCh chan<- error, doneCh <-chan int) {
+	creds, err := credsProvider.GetDockerCredentials()
+	if err != nil {
+		errCh <- err
+		return
+	}
+	ticker := time.NewTicker(credsProvider.GetRefreshTime())
+	for {
+		select {
+		case app := <-appCh:
+			log.Printf("creating secret for app %s", app)
+			err = pkg.CreateSecret(client.Secrets(app), creds)
+			if err != nil {
+				errCh <- err
+				return
+			}
+		case <-ticker.C:
+			creds, err = credsProvider.GetDockerCredentials()
+			if err != nil {
+				errCh <- err
+				return
+			}
+			for _, app := range appList {
+				log.Printf("updating secret for app %s", app)
+				err = pkg.UpdateSecret(client.Secrets(app), creds)
+				if err != nil {
+					errCh <- err
+					return
+				}
+			}
+		case <-doneCh:
+			return
+		}
+	}
+}
+
+func main() {
+	kubeClient, err := kcl.NewInCluster()
+	if err != nil {
+		log.Fatal("Error getting kubernetes client ", err)
+	}
+	params, err := pkg.GetRegistryParams()
+	if err != nil {
+		log.Fatal("Error getting registry location credentials details", err)
+	}
+	credProvider, err := credentials.GetDockerCredentialsProvider(registryLocation, params)
+	if err != nil {
+		log.Fatal("Error getting credentials provider", err)
+	}
+
+	appAddedCh := make(chan string)
+	tokenRefErrCh := make(chan error)
+	doneCh := make(chan int)
+	defer close(doneCh)
+
+	go tokenRefresher(kubeClient, credProvider, appAddedCh, tokenRefErrCh, doneCh)
+
+	for {
+		select {
+		case err = <-tokenRefErrCh:
+			log.Fatalf("error during token refresh %s", err)
+		default:
+			labelMap := labels.Set{"heritage": "deis"}
+			nsList, err := kubeClient.Namespaces().List(api.ListOptions{LabelSelector: labelMap.AsSelector(), FieldSelector: fields.Everything()})
```

Comment:
a `ns getting created, deleted and created again` within the refresh time is not an issue here as creating/updating of secrets will take care of that